### PR TITLE
[0.4.x] libvisual: Fix splitting of interleaved stereo

### PR DESCRIPTION
**Pull request by hartwork**
Previously, twice as many bytes as available were read and written (heap-buffer-overflow) and the code was not robust towards and odd number of samples.  Thanks to [AddressSanitizer](https://clang.llvm.org/docs/AddressSanitizer.html)!

### CODE REVIEW

File: libvisual/libvisual/lv_audio.c [hunk 1]
```diff
@@ -1336,12 +1344,6 @@ static int input_interleaved_stereo (VisAudioSamplePool *samplepool, VisBuffer *
 	else
 		return -1;
 
-	visual_log_return_val_if_fail (chan1 != NULL, -1);
-	visual_log_return_val_if_fail (chan2 != NULL, -1);
-
-	visual_buffer_set_destroyer (chan1, visual_buffer_destroyer_free);
-	visual_buffer_set_destroyer (chan2, visual_buffer_destroyer_free);
```
Comment (grandchild) [inline libvisual/libvisual/lv_audio.c:1343]: why are these hooks removed? the commit message does not say.

**Review by grandchild — CHANGES_REQUESTED** [review]
a small typo and extra space in the commit message:

```diff
-odd number of samples.  Thanks to AddressSanatizer!
+odd number of samples. Thanks to AddressSanitizer!
```
(but maybe that space is intentional?)

File: libvisual/libvisual/lv_audio.c (hunk 1 above)
Comment (hartwork) [reply libvisual/libvisual/lv_audio.c:1343]: They were duplicate, the `STEREO_INTERLEAVED` macro body already sets the destroyer to `visual_buffer_destroyer_free` in e.g. `chan1 = visual_buffer_new_allocate (bytes_needed_per_channel, visual_buffer_destroyer_free);` the last parameter. Good question though, I will extend the commit message accordingly :+1:

File: libvisual/libvisual/lv_audio.c [hunk 2]
```diff
@@ -1294,14 +1294,22 @@ static int sample_size_func (VisRingBuffer *ringbuffer, VisRingBufferEntry *entr
 /*  functions */
 #define STEREO_INTERLEAVED(x)											\
 		{												\
-			chan1 = visual_buffer_new_allocate (sizeof (x) * (visual_buffer_get_size (buffer) / 2),	\
+			/* Drop last sample if odd number */							\
```
[the remark this answers is not in the document]
Comment (hartwork) [reply libvisual/libvisual/lv_audio.c:1297]: Good idea, done :+1:

**Comment by hartwork** [comment]
> a small typo and extra space in the commit message:
> 
> ```diff
> -odd number of samples.  Thanks to AddressSanatizer!
> +odd number of samples. Thanks to AddressSanitizer!
> ```
> 
> (but maybe that space is intentional?)

@grandchild I keep looking up the spelling of sanitizer and still it's hard to get right, good catch! :+1: 
The space was intentional (…) but I have dropped it here now. 

Thanks for the review! :pray:

**Review by grandchild — APPROVED** [review]
mnemonic: "same as 'sanitäter' in german" 😉

**Pull request merged**